Bump k8s libraries to 1.15.3 levels

liggitt opened the pull request:
# What does this PR achieve? Why do we need it?

* Drops go1.9 support (current versions of client-go use strings.Builder, which is go1.10+... only go1.12 and go1.13 are still in support)
* Updates kubernetes client libraries from Kubernetes v1.6 levels (!) to current versions (v1.15.3)
* Removes dependency on k8s.io/kubernetes (greatly reducing vendor size from 197MB on master to 109MB with this PR)
* Needed to unblock Kubernetes moving to go1.13 (https://github.com/kubernetes/kubernetes/issues/82531)
* Added ability to specify known_hosts files for ssh exec with $SSH_KNOWN_HOSTS (previously host verification was just skipped)

### Does this PR fix issues?

Fixes #1648
Fixes #1110

### Notes for the reviewer

Updated version of https://github.com/heketi/heketi/pull/1542

cc @munnerz

nixpanic commented:
Initial failed run [#2232](https://ci.centos.org/job/gluster_heketi-functional/2232/)
Re-triggered 2nd run [#2233](https://ci.centos.org/job/gluster_heketi-functional/2233)

phlogistonjohn commented:
The new dependencies / package versions appear to be breaking the ssh executor:
```
04:08:17 [cmdexec] INFO 2019/09/13 09:08:17 Check Glusterd service status in node 192.168.10.100
04:08:17 [cmdexec] WARNING 2019/09/13 09:08:17 Failed to create SSH connection to 192.168.10.100:22: ssh: must specify HostKeyCallback
04:08:17 [cmdexec] ERROR 2019/09/13 09:08:17 heketi/executors/cmdexec/peer.go:80:cmdexec.(*CmdExecutor).GlusterdCheck: ssh: must specify HostKeyCallback
04:08:17 [heketi] ERROR 2019/09/13 09:08:17 heketi/apps/glusterfs/app_node.go:107:glusterfs.(*App).NodeAdd: ssh: must specify HostKeyCallback
04:08:17 [heketi] ERROR 2019/09/13 09:08:17 heketi/apps/glusterfs/app_node.go:108:glusterfs.(*App).NodeAdd: New Node doesn't have glusterd running
```

phlogistonjohn commented:
Searching for that error yields:
https://stackoverflow.com/a/44269276

liggitt commented:
ah, and this bumped the golang.org/x/crypto dependency and picked up https://github.com/golang/go/issues/19767

added ability to specify a list of known hosts files via $SSH_KNOWN_HOSTS, falling back to previous behavior by default

liggitt commented:
CI is happy now

humblec commented:
@liggitt cool :).. LGTM .

phlogistonjohn commented:
Before we can merge this we need to fix our "fromsource" base images to use something newer than fedora 27 which is using go 1.9, as this pr forces us to drop go 1.9. (for context: The only reason we were sticking with f27 was some bugs in newer builds in glide.) I don't want our dockerhub builds to start breaking immediately after we merge.

liggitt commented:
Thanks for the pointers. The linked issues have been fixed in glide. Removed the pin to fedora 27 and temporarily pointed the build image at this fork/branch in https://github.com/liggitt/heketi/commits/bump-k8s-go to make sure the build worked:

```
~/go/src/github.com/heketi/heketi/extras/docker/fromsource$ docker build .
...
Step 8/13 : ENV HEKETI_REPO="https://github.com/liggitt/heketi.git"
 ---> Running in f4549a3e1d7a
Removing intermediate container f4549a3e1d7a
 ---> cb4445d63041
Step 9/13 : ENV HEKETI_BRANCH="bump-k8s"
 ---> Running in 98141febb471
Removing intermediate container 98141febb471
...
 glide                       x86_64  0.13.2-1.fc30              @fedora    22 M
 golang                      x86_64  1.12.9-1.fc30              @updates  6.9 M
...
Cloning into 'heketi'...
[INFO]	Downloading dependencies. Please wait...
[INFO]	--> Fetching github.com/asaskevich/govalidator
...
[INFO]	Replacing existing vendor dependencies
[INFO]	Removing nested vendor and Godeps/_workspace directories...
go build  -ldflags "-X main.HEKETI_VERSION=v9.0.0-122-g4762ba1c-bump-k8s -extldflags '-z relro -z now'" -o heketi
make[1]: Entering directory '/build/golang/src/github.com/heketi/heketi/client/cli/go'
go build  -ldflags "-X main.HEKETI_CLI_VERSION=v9.0.0-122-g4762ba1c-bump-k8s -extldflags '-z relro -z now'" -o heketi-cli
make[1]: Leaving directory '/build/golang/src/github.com/heketi/heketi/client/cli/go'
install -D -p -m 0755 client/cli/go/heketi-cli \
	/usr/bin/heketi-cli
install -D -p -m 0755 heketi \
	/usr/bin/heketi
install -D -p -m 0644 docs/man/heketi-cli.8 \
	/usr/share/man/man8/heketi-cli.8
install -D -p -m 0755 extras/container/heketi-start.sh \
	/usr/share/heketi/container/heketi-start.sh
install -D -p -m 0644 extras/container/heketi.json \
	/usr/share/heketi/container/heketi.json
[INFO]	Glide cache has been cleared.
Dependencies resolved.
...
Successfully built ba55094f0c51
```

Updated this PR with a commit dropping the pin to fedora 27

liggitt commented:
CI is green again :)

phlogistonjohn commented:
Sounds good to me, thanks for testing out the build too!
@raghavendra-talur PTAL

liggitt commented:
build passed at https://hub.docker.com/r/heketi/heketi/builds:

Status | Tag | Commit Source | Created | Last Updated
-- | -- | -- | -- | --
Success | dev | c2e2a4a | an hour ago | 27 minutes ago

liggitt commented:
Thanks. Is there an existing timeframe for tagging a stable release that k8s can pull in to [unblock moving to go1.13](https://github.com/kubernetes/kubernetes/issues/82531)? I didn't see an existing roadmap or milestone.

liggitt commented:
> I didn't see an existing roadmap or milestone.

I take it back. [Milestone 10](https://github.com/heketi/heketi/milestone/26) was targeted two months ago :)
